google/cloud/internal/openssl_util.cc

```diff
@@ -31,6 +40,118 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace internal {
 
 namespace {
+#ifdef _WIN32
+std::string CaptureWin32Errors() {
+  auto last_error = GetLastError();
+  LPSTR message_buffer_raw = nullptr;
+  auto size = FormatMessageA(
+      FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM |
+          FORMAT_MESSAGE_IGNORE_INSERTS,
+      nullptr, last_error, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+      (LPSTR)&message_buffer_raw, 0, nullptr);
+  std::unique_ptr<char, decltype(&LocalFree)> message_buffer(message_buffer_raw,
+                                                             &LocalFree);
+  return std::string(message_buffer.get(), size) + " (error code " +
+         std::to_string(last_error) + ")";
```

Comment:
The WIN32 API is just horrible sometimes.  I had to look it up, because this cast just looks wrong, but nope, it is expected:

https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-formatmessagea

Given that the API also says:


> This buffer cannot be larger than 64K bytes.

Maybe we can simplif the code to use:

```suggestion
  auto last_error = GetLastError();
  std::vector buffer(64 * 1024, '\0');
  auto size = FormatMessageA(
      FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM |
          FORMAT_MESSAGE_IGNORE_INSERTS,
          nullptr, last_error, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
          reinterpreset_cast<LPSTR>(buffer.data(), buffer.size(), nullptr);
  return absl::StrCat(absl::string_view(buffer.data(), size), " (error code ", last_error, ")");
```

Reply:
I can do that, but it will needlessly allocate a 64KB buffer, which might be OK since this function gets called only on errors.